src/pykeen/nn/node_piece/tokenization.py

```diff
@@ -150,6 +151,76 @@ def __call__(
         # convert to torch
         return len(anchors) + 1, torch.as_tensor(tokens, dtype=torch.long)
 
+    # docstr-coverage: inherited
+    def __call__(
+        self,
+        mapped_triples: MappedTriples,
+        num_tokens: int,
+        num_entities: int,
+        num_relations: int,
+    ) -> Tuple[int, torch.LongTensor]:  # noqa: D102
+        return self._call(
+            edge_index=get_edge_index(mapped_triples=mapped_triples),
+            num_tokens=num_tokens,
+            num_entities=num_entities,
+        )
+
+
+class MetisAnchorTokenizer(AnchorTokenizer):
+    """
+    An anchor tokenizer, which first partitions the graph using Metis.
+
+    We use the binding by `torch-sparse`. The Metis graph partitioning algorithm is described here:
+    http://glaros.dtc.umn.edu/gkhome/metis/metis/overview
+    """
+
+    def __init__(self, num_partitions: int = 2, **kwargs):
+        """Initialize the tokenizer.
+
+        :param num_partitions:
+            the number of partitions obtained through Metis.
+        :param kwargs:
+            additional keyword-based parameters passed to :meth:`AnchorTokenizer.__init__`. note that there will be one
+            anchor tokenizer per partition, i.e., the vocabulary size will grow respectively.
+        """
+        super().__init__(**kwargs)
+        self.num_partitions = num_partitions
+
+    # docstr-coverage: inherited
+    def __call__(
+        self,
+        mapped_triples: MappedTriples,
+        num_tokens: int,
+        num_entities: int,
+        num_relations: int,
+    ) -> Tuple[int, torch.LongTensor]:  # noqa: D102
+        try:
+            import torch_sparse
+        except ImportError as err:
+            raise ImportError(f"{self.__class__.__name__} requires `torch_sparse` to be installed.") from err
+
+        row, col = get_edge_index(mapped_triples=mapped_triples)
+        re_ordered_adjacency, bound, perm = torch_sparse.partition(
+            src=torch_sparse.SparseTensor(row=row, col=col), num_parts=self.num_partitions
+        )
+        # select independently per partition
+        vocabulary_size = 0
+        assignment = []
+        for low, high in more_itertools.pairwise(bound.tolist()):
+            # select adjacency part;
+            # note: the indices will automatically be in [0, ..., high - low), since they are *local* indices
+            edge_index = re_ordered_adjacency[low:high, low:high].to_torch_sparse_coo_tensor().coalesce().indices()
+            this_vocabulary_size, this_assignment = super(self.__class__, self)._call(
+                edge_index=edge_index, num_tokens=num_tokens, num_entities=high - low
+            )
+            # offset
+            this_assignment = this_assignment + vocabulary_size
+            # note: permutation will be later on reverted
+            vocabulary_size += this_vocabulary_size
+            assignment.append(this_assignment)
```

Comment:
Ideally, we could replace this for-loop with something parallel (joblib?) as those tokenizations are independent from each other

Reply:
sounds like the beginning of a massively complicated effort to parallelize this in many different ways ;)

Reply:
(OT: interesting node/edge ratio)

Reply:
btw partitioning works on a GPU, too, thanks to the CUDA implementation - for me the whole Wikidata5M on a GPU took less than a minute with 10 partitions

Reply:
conceptnet, which is afaik the largest one we have easily available, runs in under 5 min, where less than one minute is spent with the super class, and the rest for calculating the partition :sweat_smile: 

I guess we can ignore parallelizing the partition-wise tokenization for now.

Reply:
That's fast for the 28M ConceptNet 👀  What is the total number of anchors / anchors per node you mined for this graph?

Reply:
64 partitions, 32 anchors per partition

Reply:
Running on ConceptNet (28M nodes) with 64 partitions / 10k anchors per each, METIS took about 2 minutes to complete, now BFS tokenization of each ~450K nodes partition with 10k anchors takes 5 minutes each 😉 64 x 5 would be 5 hours